src/js/node/net.ts

```diff
@@ -469,11 +497,431 @@ const ServerHandlers: SocketHandler = {
+    const { self, that } = socket.data;
+    if (self[kended]) return;
+    self[kended] = true;
+    if (!self.allowHalfOpen) self.write = writeAfterFIN;
+    self.push(null);
+    self.read(0);
+  },
+  close(socket, err) {
+    $debug("Bun.Socket close");
+    let { self, that } = socket.data;
+    if (!that) that = SocketHandle[kAttach](socket, self);
+    if (self[kclosed]) return;
+    self[kclosed] = true;
+    self[kended] = true;
+    if (!self.allowHalfOpen) self.write = writeAfterFIN;
+    self.push(null);
+    self.read(0);
+  },
+  handshake(socket, success, verifyError) {
+    $debug("Bun.Socket handshake");
+    const { self, that } = socket.data;
+    if (!success && verifyError?.code === "ECONNRESET") {
+      // will be handled in onConnectEnd
+      return;
+    }
+
+    self._securePending = false;
+    self.secureConnecting = false;
+    self._secureEstablished = !!success;
+
+    self.emit("secure", self);
+    self.alpnProtocol = socket.alpnProtocol;
+    const { checkServerIdentity } = self[bunTLSConnectOptions];
+    if (!verifyError && typeof checkServerIdentity === "function" && self.servername) {
+      const cert = self.getPeerCertificate(true);
+      verifyError = checkServerIdentity(self.servername, cert);
+    }
+    if (self._requestCert || self._rejectUnauthorized) {
+      if (verifyError) {
+        self.authorized = false;
+        self.authorizationError = verifyError.code || verifyError.message;
+        if (self._rejectUnauthorized) {
+          self.destroy(verifyError);
+          return;
+        }
+      } else {
+        self.authorized = true;
+      }
+    } else {
+      self.authorized = true;
+    }
+    self.emit("secureConnect", verifyError);
+    self.removeListener("end", onConnectEnd);
+  },
+  error(socket, error, ignoreHadError) {
+    $debug("Bun.Socket error");
+    const { self, that } = socket.data;
+    if (self._hadError && !ignoreHadError) return;
+    self._hadError = true;
+
+    const callback = self[kwriteCallback];
+    if (callback) {
+      self[kwriteCallback] = null;
+      callback(error);
+    }
+    self.emit("error", error);
+
+    if (!self.destroyed) process.nextTick(destroyNT, self, error);
+  },
+  timeout(socket) {
+    $debug("Bun.Socket timeout");
+    const { self, that } = socket.data;
+    self.emit("timeout", self);
+  },
+  connectError(socket, error) {
+    $debug("Bun.Socket connectError");
+    let { self, that, req } = socket.data;
+    if (!that) that = SocketHandle[kAttach](socket, self);
+    self._handle = that;
+    socket[owner_symbol] = self;
+    that[ksocket] = socket;
+    that[kpromise] = null;
+    req!.oncomplete(error.errno, self._handle, req, true, true);
+    socket.data.req = undefined;
+  },
 };
 
+const kConnectTcp = Symbol("kConnectTcp");
+const kConnectPipe = Symbol("kConnectPipe");
+
+class SocketHandle {
+  #promise: Promise<Socket<undefined>> | null;
+  #socket: Socket<undefined> | null;
+  #refsOnConnect: number = 0;
```

Comment:
This ref counting doesn't do what you probably were trying to do because ref and unref are booleans. So the last one always wins.